apollo-router/src/plugins/authentication.rs

```diff
@@ -0,0 +1,901 @@
+//! Authentication plugin
+// With regards to ELv2 licensing, this entire file is license key functionality
+
+use std::ops::ControlFlow;
+use std::str::FromStr;
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::time::Duration;
+
+use deduplicate::Deduplicate;
+use deduplicate::DeduplicateFuture;
+use http::header::ACCEPT;
+use http::header::CONTENT_TYPE;
+use http::StatusCode;
+use jsonwebtoken::decode;
+use jsonwebtoken::decode_header;
+use jsonwebtoken::jwk::JwkSet;
+use jsonwebtoken::DecodingKey;
+use jsonwebtoken::Validation;
+use once_cell::sync::Lazy;
+use reqwest::Client;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use tokio::fs::read_to_string;
+use tower::BoxError;
+use tower::ServiceBuilder;
+use tower::ServiceExt;
+use url::Url;
+
+use crate::error::LicenseError;
+use crate::graphql;
+use crate::layers::ServiceBuilderExt;
+use crate::plugin::Plugin;
+use crate::plugin::PluginInit;
+use crate::register_plugin;
+use crate::services::apollo_graph_reference;
+use crate::services::router;
+use crate::Context;
+
+type SharedDeduplicate = Arc<
+    Deduplicate<Box<dyn Fn(Url) -> DeduplicateFuture<JwkSet> + Send + Sync + 'static>, Url, JwkSet>,
+>;
+
+pub(crate) const AUTHENTICATION_SPAN_NAME: &str = "authentication plugin";
+
+const DEFAULT_AUTHENTICATION_NETWORK_TIMEOUT: Duration = Duration::from_secs(15);
+
+const DEFAULT_AUTHENTICATION_COOLDOWN: Duration = Duration::from_secs(15);
+
+static COOLDOWN: Lazy<AtomicBool> = Lazy::new(|| AtomicBool::new(false));
+
+static CLIENT: Lazy<Result<Client, BoxError>> = Lazy::new(|| {
+    apollo_graph_reference().ok_or(LicenseError::MissingGraphReference)?;
+    Ok(Client::new())
+});
+
+struct AuthenticationPlugin {
+    configuration: JWTConf,
+    jwks: SharedDeduplicate,
+    jwks_url: Url,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, JsonSchema)]
+struct JWTConf {
+    /// Retrieve our JWK Set from here
+    jwks_url: String,
+    /// HTTP header expected to contain JWT
+    #[serde(default = "default_header_name")]
+    header_name: String,
+    /// Header prefix
+    #[serde(default = "default_header_prefix")]
+    header_prefix: String,
```

Comment:
```suggestion
    header_value_prefix: String,
```
to avoid confusion with the header name ?

Reply:
done: b8c48ba5cd7bdb418b3b0834ecf333c41479de4c